File: src/memory/retrieval/session.ts
/**
 * Session-start lesson retrieval
 *
 * Loads high-severity lessons at the start of a session.
 * No vector search - just filter by severity and recency.
 */

import { incrementRetrievalCount, readMemoryItems } from '../storage/index.js';
import type { MemoryItem, Severity } from '../types.js';

/** Default number of lessons to load at session start */
const DEFAULT_LIMIT = 5;

/** A memory item with severity field present */
type LessonWithSeverity = MemoryItem & { severity: Severity };

/**
 * Type guard to check if a memory item has severity set
 */
function hasSeverity(item: MemoryItem): item is MemoryItem & { severity: Severity } {
  return item.severity !== undefined;
}

/**
 * Load high-severity lessons for session start.
 *
 * Returns confirmed, high-severity lessons sorted by recency.
 * These are the most important lessons to surface at the start
 * of a coding session.
 *
 * @param repoRoot - Repository root directory
 * @param limit - Maximum number of lessons to return (default: 5)
 * @returns Array of high-severity lessons, most recent first
 */
export async function loadSessionLessons(
  repoRoot: string,
  limit: number = DEFAULT_LIMIT
): Promise<LessonWithSeverity[]> {
  const { items } = await readMemoryItems(repoRoot);

  // Filter for high-severity, confirmed items of any type (excluding invalidated)
  const highSeverityLessons = items.filter(
    (item): item is MemoryItem & { severity: Severity } =>
      hasSeverity(item) &&
      item.severity === 'high' &&
      item.confirmed &&
      !item.invalidatedAt
  );

  // Sort by recency (most recent first)
  highSeverityLessons.sort((a, b) => {
    const dateA = new Date(a.created).getTime();
    const dateB = new Date(b.created).getTime();
    return dateB - dateA;
  });

  // Return top N and track surfaced lessons as retrieved.
  const topLessons = highSeverityLessons.slice(0, limit);
  if (topLessons.length > 0) {
    incrementRetrievalCount(repoRoot, topLessons.map((lesson) => lesson.id));
  }

  return topLessons;
}

/**
 * @deprecated Use loadSessionMemory. Backward-compat alias.
 */
export const loadSessionMemory = loadSessionLessons;

File: go/internal/retrieval/session.go
package retrieval

import (
	"sort"
	"time"

	"github.com/nathandelacretaz/compound-agent/internal/memory"
)

// DefaultSessionLimit is the default number of session lessons to return.
const DefaultSessionLimit = 5

// LoadSessionLessons returns high-severity, confirmed, non-invalidated lessons
// sorted by recency (most recent first), limited to `limit`.
func LoadSessionLessons(repoRoot string, limit int) ([]memory.MemoryItem, error) {
	if limit <= 0 {
		limit = DefaultSessionLimit
	}

	result, err := memory.ReadMemoryItems(repoRoot)
	if err != nil {
		return nil, err
	}

	var filtered []memory.MemoryItem
	for _, item := range result.Items {
		if item.Severity == nil || *item.Severity != memory.SeverityHigh {
			continue
		}
		if !item.Confirmed {
			continue
		}
		if item.InvalidatedAt != nil {
			continue
		}
		filtered = append(filtered, item)
	}

	// Sort by Created descending (most recent first)
	sort.Slice(filtered, func(i, j int) bool {
		ti, errI := time.Parse(time.RFC3339, filtered[i].Created)
		tj, errJ := time.Parse(time.RFC3339, filtered[j].Created)
		if errI != nil || errJ != nil {
			return filtered[i].Created > filtered[j].Created
		}
		return ti.After(tj)
	})

	if len(filtered) > limit {
		filtered = filtered[:limit]
	}

	return filtered, nil
}
